### PR TITLE
add patchwork

### LINKED ISSUE #59100: add patchbay (secure scuttlebutt)
## Issue description

i'd like to use secure scuttlebutt client patchbay and/or patchwork

## Technical details

Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the
results.

### PR DESCRIPTION
**Pull request by thedavidmeister**
fixes #59100

###### Motivation for this change

Adds patchwork, official client for secure scuttlebutt.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Assured whether relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Comment by thedavidmeister** [comment]
i kind of bumbled through this by copying the `joplin-desktop` and `standardnotes` derivations

the `gsettings-desktop-schemas` seems to be needed to prevent crashes when uploading profile pics

**Comment by thedavidmeister** [comment]
seems to run ok but there are warnings:

this on boot: 

```
(process:20493): Gtk-WARNING **: 22:25:30.050: Locale not supported by C library.
        Using the fallback 'C' locale.
{ path: '/home/thedavidmeister/.ssb',
  party: true,
  timeout: 0,
  pub: true,
  local: true,
  friends: { dunbar: 150, hops: 2 },
  gossip: { connections: 3 },
  connections: 
   { outgoing: { net: [Array] },
     incoming: { net: [Array], ws: [Array] } },
  timers: { connection: 0, reconnect: 5000, ping: 300000, handshake: 5000 },
  caps: 
   { shs: '1KHLiKZvAvjbY1ziZEHMXawbCEIM6qwjCDm3VYRan/s=',
     sign: null },
  master: [],
  logging: { level: 'notice' },
  port: 8008,
  blobsPort: 8989,
  server: true,
  _: [],
  host: '::',
  ws: { host: '::1', port: 8989, scope: [ 'device' ], transform: 'shs' },
  keys: 
   { curve: 'ed25519',
     public: '8VYZW+U0zbewBQSaKpi1fCTD2cPEbkKy8B6amcy06IY=.ed25519',
     private: null,
     id: '@8VYZW+U0zbewBQSaKpi1fCTD2cPEbkKy8B6amcy06IY=.ed25519' } }
ssb-friends: createFriendStream legacy api used
git-ssb is not installed, or not available in path
Error: listen EADDRINUSE :::8989
    at Object._errnoException (util.js:1024:11)
    at _exceptionWithHostPort (util.js:1046:20)
    at Server.setupListenHandle [as _listen2] (net.js:1351:14)
    at listenInCluster (net.js:1392:12)
    at Server.listen (net.js:1476:7)
    at EventEmitter.emitter.listen (/home/thedavidmeister/.cache/appimage-run/9811e63b0395fab7bf604f9dcd63c60ad91457d1378881365e26ae2aa8c159b4/squashfs-root/app/resources/app.asar/node_modules/pull-ws/server.js:50:14)
    at Object.server (/home/thedavidmeister/.cache/appimage-run/9811e63b0395fab7bf604f9dcd63c60ad91457d1378881365e26ae2aa8c159b4/squashfs-root/app/resources/app.asar/node_modules/scuttlebot/node_modules/multiserver/plugins/ws.js:20:31)
    at Object.server (/home/thedavidmeister/.cache/appimage-run/9811e63b0395fab7bf604f9dcd63c60ad91457d1378881365e26ae2aa8c159b4/squashfs-root/app/resources/app.asar/node_modules/multiserver/compose.js:96:29)
    at /home/thedavidmeister/.cache/appimage-run/9811e63b0395fab7bf604f9dcd63c60ad91457d1378881365e26ae2aa8c159b4/squashfs-root/app/resources/app.asar/node_modules/multiserver/index.js:45:21
    at Array.map (<anonymous>)
ssb-friends: stream legacy api used
ssb-friends: get legacy api used
```

then a lot of this:

```
(ssb-patchwork:20493): dconf-WARNING **: 22:27:31.414: failed to commit changes to dconf: GDBus.Error:org.freedesktop.DBus.Error.ServiceUnknown: The name ca.desrt.dconf was not provided by any .service files

(ssb-patchwork:20493): dconf-WARNING **: 22:27:31.431: failed to commit changes to dconf: GDBus.Error:org.freedesktop.DBus.Error.ServiceUnknown: The name ca.desrt.dconf was not provided by any .service files

(ssb-patchwork:20493): dconf-WARNING **: 22:27:31.445: failed to commit changes to dconf: GDBus.Error:org.freedesktop.DBus.Error.ServiceUnknown: The name ca.desrt.dconf was not provided by any .service files

(ssb-patchwork:20493): dconf-WARNING **: 22:27:31.446: failed to commit changes to dconf: GDBus.Error:org.freedesktop.DBus.Error.ServiceUnknown: The name ca.desrt.dconf was not provided by any .service files

```

File: pkgs/applications/social/patchwork/default.nix
Comment (bricewge) [inline pkgs/applications/social/patchwork/default.nix]: I think you should use `appimageTools`, take a look at #54693.

File: pkgs/applications/social/patchwork/default.nix
Comment (bricewge) [inline pkgs/applications/social/patchwork/default.nix]: From your log I think you are missing `git-ssb` from `nodePackages`.

**Review by bricewge — CHANGES_REQUESTED** [review]
As discussed in https://github.com/NixOS/nixpkgs/pull/53156#discussion_r244649350 you should probably renamed the package and put it in `pkgs/applications/networking/ssb/` alongside `patchwork-classic`.

Related with your bug report #59100

**Comment by christianbundy** [comment]
@thedavidmeister I think the error you're seeing is coming from https://github.com/ssbc/patchwork/issues/972

File: pkgs/applications/social/patchwork/default.nix
Comment (samueldr) [inline pkgs/applications/social/patchwork/default.nix]: You will need to add yourself to `maintainers/maintainer-list.nix`; this is why "This PR breaks listing of package outputs after merging".

File: pkgs/applications/social/patchwork/default.nix
Comment (christianbundy) [reply pkgs/applications/social/patchwork/default.nix]: It looks like an error, but `git-ssb` is optional and the error is only being output for debugging. I'll prioritize fixing that debug output so that it's more obvious.

**Comment by matthiasbeyer** [comment]
Is this "patchwork" or "patchwork-classic"? Because if it is the latter, don't mind continuing... patchwork-classic is afaik deprecated and only patchwork should be used.

If it is patchwork, I'd welcome to have this in nixpkgs!

**Comment by thedavidmeister** [comment]
@matthiasbeyer patchwork, i got the latest appimage

File: pkgs/applications/social/patchwork/default.nix
Comment (thedavidmeister) [reply pkgs/applications/social/patchwork/default.nix]: @bricewge i've mirrored other examples so i don't think this is needed

https://github.com/NixOS/nixpkgs/pull/54693 my understanding of the conversation here is that `appimageTools` referenced in that PR is that it is for the build side, not the download and run side

**Comment by thedavidmeister** [comment]
i tested `3.11.4` with `nix-env -f . -iA patchwork`

running this version has a lot fewer errors than newer version(s): https://github.com/ssbc/patchwork/issues/972

tried adding `git-ssb` but it didn't help (build time vs. run time path differences?), i don't think it is needed as it is debug info (check comment thread)

moved the derivation to `networking/ssb/patchwork`

fixed `maintainer-list.nix`

don't think `appimageTools` is correct here as we are consuming not building an app image

File: pkgs/applications/networking/ssb/patchwork/default.nix
Comment (asymmetric) [inline pkgs/applications/networking/ssb/patchwork/default.nix]: I think the new best practice is to just use `pname = "patchwork"`

File: pkgs/applications/networking/ssb/patchwork/default.nix
Comment (asymmetric) [inline pkgs/applications/networking/ssb/patchwork/default.nix]: Can we split this? It's a bit long.

File: pkgs/applications/networking/ssb/patchwork/default.nix
Comment (asymmetric) [inline pkgs/applications/networking/ssb/patchwork/default.nix]: Have you tried 3.11.5? The issue mentions 3.11.6 as being broken.

File: pkgs/applications/networking/ssb/patchwork/default.nix
Comment (asymmetric) [inline pkgs/applications/networking/ssb/patchwork/default.nix]: You could use `pname` here.

File: pkgs/applications/networking/ssb/patchwork/default.nix
Comment (asymmetric) [inline pkgs/applications/networking/ssb/patchwork/default.nix]: How come only `x86_64-linux` and not e.g. `platforms.linux` or `platforms.unix`?

File: pkgs/applications/networking/ssb/patchwork/default.nix
Comment (thedavidmeister) [reply pkgs/applications/networking/ssb/patchwork/default.nix]: @asymmetric the issue mentions both 3.11.6 and 3.11.5 as broken https://github.com/ssbc/patchwork/issues/972#issuecomment-476215494

File: pkgs/applications/networking/ssb/patchwork/default.nix
Comment (thedavidmeister) [reply pkgs/applications/networking/ssb/patchwork/default.nix]: @asymmetric i don't know, i copied this from elsewhere `platforms.linux` would make sense i think

File: pkgs/applications/networking/ssb/patchwork/default.nix
Comment (thedavidmeister) [reply pkgs/applications/networking/ssb/patchwork/default.nix]: @asymmetric is there some kind of formatting tool i can run?

File: pkgs/applications/networking/ssb/patchwork/default.nix
Comment (christianbundy) [reply pkgs/applications/networking/ssb/patchwork/default.nix]: Yeah, sorry about that. We'll have a new release out soon, currently just blocked by https://github.com/ssbc/patchwork/issues/978.

Comment (thedavidmeister) [reply pkgs/applications/networking/ssb/patchwork/default.nix]: i think that's a followup

i tried the new release alphas and they didn't even load - we can always tweak this later

Comment (christianbundy) [reply pkgs/applications/networking/ssb/patchwork/default.nix]: That's no good, sorry to hear that. If you have time to make an issue, I'd love to help debug and ensure that it works on your system. :+1:

File: pkgs/applications/networking/ssb/patchwork/default.nix
Comment (asymmetric) [reply pkgs/applications/networking/ssb/patchwork/default.nix]: https://discourse.nixos.org/t/job-implement-a-nix-fmt-formatter/2819

File: pkgs/applications/networking/ssb/patchwork/default.nix
Comment (asymmetric) [reply pkgs/applications/networking/ssb/patchwork/default.nix]: I would try with the most general possible and see if Hydra/ofBorg tests fail, and then shrink the scope if they do.

Not sure this is the correct approach and maybe `with platforms; [ linux darwin ]` is enough.

Comment (flokli) [reply pkgs/applications/networking/ssb/patchwork/default.nix]: Looking at https://github.com/ssbc/patchwork/releases/, upstream only publishes x86_64 appimages. So if we don't compile from source, this will only work on "x86_64-linux".

Comment (asymmetric) [reply pkgs/applications/networking/ssb/patchwork/default.nix]: Of course - AppImage is Linux only 🤦‍♂

File: pkgs/applications/networking/ssb/patchwork/default.nix
Comment (thedavidmeister) [reply pkgs/applications/networking/ssb/patchwork/default.nix]: sure thing, but 3.11.4 works fine for me, i'd rather treat debugging alphas for a future release as a followup

File: pkgs/applications/networking/ssb/patchwork/default.nix
Comment (thedavidmeister) [reply pkgs/applications/networking/ssb/patchwork/default.nix]: i did try to do node2nix and similar, and asked on #nixos IRC a few times for help, and someone else in ssb did try as well..

it's too hard for beginners to get node working

can we get the AppImage in for linux and then if someone more knowledgeable wants to get the electron app building cross-platform they can?

File: pkgs/applications/networking/ssb/patchwork/default.nix
Comment (thedavidmeister) [reply pkgs/applications/networking/ssb/patchwork/default.nix]: aight, so there's a WIP tool but nothing stable or "official"

i have zero interest in debating fmt rules outside the context of automated tooling so i'll just do what you're suggesting

File: pkgs/applications/networking/ssb/patchwork/default.nix
Comment (thedavidmeister) [reply pkgs/applications/networking/ssb/patchwork/default.nix]:
```
error: derivation name missing
```

File: pkgs/applications/networking/ssb/patchwork/default.nix
Comment (thedavidmeister) [reply pkgs/applications/networking/ssb/patchwork/default.nix]: see error above

**Comment by thedavidmeister** [comment]
thanks for the review @asymmetric i implemented what i could, how do you feel about it now?

File: pkgs/applications/networking/ssb/patchwork/default.nix
Comment (asymmetric) [reply pkgs/applications/networking/ssb/patchwork/default.nix]: You also need `inherit version;` below.

**Comment by asymmetric** [comment]
Added a comment.

**Comment by thedavidmeister** [comment]
@asymmetric i don't see it :thinking:

File: pkgs/applications/networking/ssb/patchwork/default.nix
Comment (christianbundy) [reply pkgs/applications/networking/ssb/patchwork/default.nix]: Sorry, I read "broken" and thought it was talking about the security vulnerabilities. There's a security issue in 3.11.4 so I'd recommend upgrading ASAP.

**Comment by asymmetric** [comment]
Because you defined `version` outside of the package's expression, you need to `inherit version;`, otherwise the full package name cannot be composed.

**Comment by thedavidmeister** [comment]
@asymmetric great, seems to work with the `version` on the derivation rather than the let

how you feeling about it now?

**Comment by asymmetric** [comment]
Nice!

You should clean up the commit history and make it fit [this](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md#submitting-changes).

In short, there should probably be just one commit (maybe two, to add yourself to the maintainers), and it should look like `patchwork: init at 3.11.4`.

**Comment by thedavidmeister** [comment]
@asymmetric doesn't github provide the option to squash commits when merging the PR?

**Comment by matthiasbeyer** [comment]
Yes, but it does not provide options for cleaning up the branch. You made several changes which should end up in different commits, but not several changes which should end up in one commit.

**Comment by asymmetric** [comment]
Also I think it's nice not to ask the maintainers to clean up your branch when merging, even if in theory they could :)

But I agree that there's a lot of things to check manually when contributing to nixpkgs.

**Comment by thedavidmeister** [comment]
This is the real history of the code development, rewriting it to fit someone(s) else’s aesthetics is beyond me I’m afraid

IMO “cleaning up” is a git smell, I don’t expect anyone to ever do it! but I’ll roll with this idea to get a merge approved

Would 1 commit for maintainers and 1 for everything else as suggested earlier provide sufficient dopamine for everyone? 😉

**Comment by samueldr** [comment]
Yes two commits is fine. One for adding yourself to the maintainers list, and the other one for the package init is what would be expected. And thanks for the contribution!

* * *

The [Making patches](https://nixos.org/nixpkgs/manual/#submitting-changes-making-patches) section of the manual has the exact format for packages and modules.

The manual explains it as “Make commits of logical units”. The main idea being that yes, while developing things are messy, and it's natural, they're not necessarily what's published here. It is “a smell”, but I'm not sure it's a bad smell, considering the scope of Nixpkgs, and how git, this way, becomes the history of the main changes, not the development history of separate features. This is part of [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

**Comment by thedavidmeister** [comment]
@samueldr what is the commit message for adding myself to maintainers?

**Comment by samueldr** [comment]
When in doubt, [I check the history](https://github.com/NixOS/nixpkgs/commits/master/maintainers/maintainer-list.nix) and try to reason with the general rules.

Here `maintainers: Add thedavidmeister` could be fine, as an example.

Package inits and upgrades have rules that [allows automation](https://github.com/NixOS/ofborg/#automatic-building).

**Comment by thedavidmeister** [comment]
@samueldr kk, here it is https://github.com/NixOS/nixpkgs/pull/61530

**Pull request closed without merging**
